SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/ReferenceCountedLockingPolicy.cpp

```diff
@@ -0,0 +1,82 @@
+#include "EngineClasses/SpatialPackageMapClient.h"
+#include "Interop/SpatialStaticComponentView.h"
+#include "Schema/AuthorityIntent.h"
+
+DEFINE_LOG_CATEGORY(LogReferenceCountedLockingPolicy);
+
+ActorLockToken UReferenceCountedLockingPolicy::AcquireLock(const AActor* Actor)
+{
+	check(Actor != nullptr);
+	check(CanAcquireLock(Actor))
+	++ActorToReferenceCount.FindOrAdd(Actor);
+	TokenToNameAndActor.Emplace(NextToken, LockNameAndActor{ FString(), Actor });
+	UE_LOG(LogReferenceCountedLockingPolicy, Log, TEXT("Acquiring migration lock. Actor: %s. Locks held: %d."), *Actor->GetName(), *ActorToReferenceCount.Find(Actor));
+	return NextToken++;
+}
+
+ActorLockToken UReferenceCountedLockingPolicy::AcquireLock(const AActor* Actor, FString DebugString)
+{
+	check(Actor != nullptr);
+	check(CanAcquireLock(Actor))
+	++ActorToReferenceCount.FindOrAdd(Actor);
+	UE_LOG(LogReferenceCountedLockingPolicy, Log, TEXT("Acquiring migration lock. "
+		"Actor: %s. Lock name: %s. Token %d: Locks held: %d."), *Actor->GetName(), *DebugString, NextToken, *ActorToReferenceCount.Find(Actor));
+	TokenToNameAndActor.Emplace(NextToken, LockNameAndActor{ MoveTemp(DebugString), Actor });
+	return NextToken++;
+}
+
+bool UReferenceCountedLockingPolicy::CanAcquireLock(const AActor* Actor) const
+{
+	const auto* NetDriver = Cast<USpatialNetDriver>(Actor->GetWorld()->GetNetDriver());
+	const auto EntityId = NetDriver->PackageMap->GetEntityIdFromObject(Actor);
+
+	const bool HasAuthIntent = NetDriver->VirtualWorkerTranslator->GetLocalVirtualWorkerId() ==
+			NetDriver->StaticComponentView->GetComponentData<SpatialGDK::AuthorityIntent>(EntityId)->VirtualWorkerId;
+	const bool HasAuth = NetDriver->StaticComponentView->GetAuthority(EntityId, SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID) == WORKER_AUTHORITY_AUTHORITATIVE;
+	if (!HasAuth)
+	{
+		UE_LOG(LogReferenceCountedLockingPolicy, Log, TEXT("Can not lock actor migration. Do not have authority. Actor: %s"), *Actor->GetName());
+	}
+	if (!HasAuthIntent)
+	{
+		UE_LOG(LogReferenceCountedLockingPolicy, Log, TEXT("Can not lock actor migration. Authority intent does not match this worker. Actor: %s"), *Actor->GetName());
+	}
+	return HasAuthIntent && HasAuth;
+}
+
+void UReferenceCountedLockingPolicy::ReleaseLock(ActorLockToken Token)
+{
+	const auto NameAndActor = TokenToNameAndActor.FindAndRemoveChecked(Token);
+	const AActor* Actor = NameAndActor.Actor;
+	const FString& Name = NameAndActor.LockName;
+	UE_LOG(LogReferenceCountedLockingPolicy, Log, TEXT("Releasing actor migration lock. Actor: %s. Token: %d. Lock name: %s"), *Actor->GetName(), Token, *Name);
+
+	check(ActorToReferenceCount.Contains(Actor));
+
+	{
+		// Reduce the reference count and erase the entry if reduced to 0.
+		auto CountIt = ActorToReferenceCount.CreateKeyIterator(Actor);
+		auto& Count = CountIt.Value();
+		if (Count == 1)
+		{
+			UE_LOG(LogReferenceCountedLockingPolicy, Log, TEXT("Actor migration no longer lock. Actor: %s"), *Actor->GetName(), *Name);
```

Comment:
I think, there's a chance `const AActor* Actor` might point to invalid data.

Reply:
How come?

Reply:
If Actor gets destroyed, while being locked, the stored raw Actor pointer is still there and points to invalid data.

Reply:
Talked to Sami and we concluded we should add this as a precondition that you should never call release lock outside the lifetime of the actor. Unless you can think of a way we can check this?

Reply:
I don't think there's a way to check that Actor is valid. What we could do though, is store the Actor name in the Policy, instead of getting it from Actor*.

Did you add the precondition check? If so, where can I see it?

Reply:
Is there any update on this conversation?

Reply:
Ah, found it, there's an `OnLockedActorDeleted` callback now 👍